File: arrays/04_best_time2buy_sell_stock/best_time2buy_sell_stock.py
from typing import List


class Solution:
    def max_profit(self, prices: List[int]) -> int:
        if not self._check_constraints(prices):
            raise ValueError("Constraints violated")

        min_price, max_profit = 10**4, 0
        for price in prices:
            min_price = min(min_price, price)
            max_profit = max(max_profit, price - min_price)

        return max_profit

    def _check_constraints(self, prices: List[int]) -> bool:
        n = len(prices)
        if n < 1 or n > 10**5:
            return False

        for price in prices:
            if price < 0 or price > 10**4:
                return False

        return True

File: arrays/04_best_time2buy_sell_stock/best_time2buy_sell_stock.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <stdexcept>

using namespace std;


class Solution {
public:
    int maxProfit(const vector<int>& prices) {
        if (!checkConstraints(prices)) {
            throw std::runtime_error("Constraints violated");
        }
        
        int minPrice = pow(10, 4);
        int maxProfit = 0;

        for (int i = 0; i < prices.size(); i++) {
            minPrice = min(minPrice, prices[i]);
            maxProfit = max(maxProfit, prices[i] - minPrice);
        }

        return maxProfit;
    }

private:
    // Helper function to check the constraints
    bool checkConstraints(const vector<int>& prices) {
        int n = prices.size() + 1; 
        if (n < 1 || n > pow(10, 5)) return false;

        for (auto& price : prices) {
            if (price < 0 || price > pow(10, 4)) return false;
        }

        return true;
    }
};
